fava: init at 1.0

matthiasbeyer opened the pull request:
**WIP**

It wants `flask-babel`, though adding the package `flaskbabel` (after upgrading it from 0.9 to 0.10 [which does not change the hash?]) doesn't help. The build also says that the requirement is already satisfied. :curly_loop: 
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

matthiasbeyer committed:
fava: init at 0.3.0

fpletz commented:
Just tried this but a lot more dependencies like dateutils, pygments and livereload (not packaged yet) are missing.

matthiasbeyer committed:
pythonPackages.livereload: init at 2.4.1

matthiasbeyer commented:
I added some more patches here, but now I get the following error I do not understand:

```

fetching path ‘/nix/store/p5m4wkh3qrpcjfz5q5k1gfaw53ncjd82-beancount-2016-04-10-b5721f1c6f01bd168a5781652e5e3167f7f8ceb3’...

*** Downloading ‘https://cache.nixos.org/nar/0jyl865qx8f63j6wwn92rlf51z2vjccw5zhc8kjv6q5sbfcrnbpv.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/p5m4wkh3qrpcjfz5q5k1gfaw53ncjd82-beancount-2016-04-10-b5721f1c6f01bd168a5781652e5e3167f7f8ceb3’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  722k  100  722k    0     0   216k      0  0:00:03  0:00:03 --:--:--  216k

building path(s) ‘/nix/store/r7hvwr656cf9q2xrphi3b7gfjjprklvh-fava-0.3.0’
unpacking sources
unpacking source archive /nix/store/hgajs0jhzdagrkl3k4xk9hvf3rkchi6i-v0.3.0.tar.gz
source root is fava-0.3.0
setting SOURCE_DATE_EPOCH to timestamp 1464077662 of file fava-0.3.0/tox.ini
patching sources
configuring
building
usage: nix_run_setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: nix_run_setup.py --help [cmd1 cmd2 ...]
   or: nix_run_setup.py --help-commands
   or: nix_run_setup.py cmd --help

error: invalid command 'bdist_wheel'
builder for ‘/nix/store/wf2khqd5z72ak43p34j1q8d76pv2vqgj-fava-0.3.0.drv’ failed with exit code 1
error: build of ‘/nix/store/wf2khqd5z72ak43p34j1q8d76pv2vqgj-fava-0.3.0.drv’ failed

```

Mic92 commented:
You need `wheel` as dependency.

matthiasbeyer commented:
I get the same output when adding `wheel` to `buildInputs`.

Mic92 commented:
this is a python3 only package: https://github.com/aumayr/fava/blob/master/setup.py#L66

matthiasbeyer committed:
beancount-pygments-lexer: init at 0.1.1

matthiasbeyer commented:
Another bunch of patches. The `beancount-pygments-lexer` fails now and I don't know why. I also tried to disable the checks with `doCheck = false`, but the tests run anyways.... not sure what the issue here is ...

matthiasbeyer commented:
Closed by accident, sorry.

matthiasbeyer commented:
Okay, it builds now. @Mic92 could you help me with the following issue: fava does not yet find the static stuff, like CSS and JS which are required to run the website properly.

To reproduce my issue:

```bash
cat <<EOS > example.beancount
2016-01-01 open Assets:Cash
EOS
nix-build -A fava
./result/bin/fava ./example.beancount
```

And then go to `localhost:5000` - see the output of fava, it yields 404 for all static things and so on ... I don't know how to properly solve this issue.

Mic92 commented:
those files seems to be not included in the package:

```
$ cd /nix/store/6873wpyyz4rxj3m2w1s91x7gjh8hhvds-fava-0.3.0/lib/python3.5/site-packages/fava
$ find static
static
static/images
static/images/favicon.ico
```

matthiasbeyer commented:
@Mic92 After copying them with 

```nix
postInstall = ''
  cp fava/static -r $out/lib/python3.5/site-packages/fava/
'';
```

It doesn't work either.

matthiasbeyer commented:
Maybe you, @FRidh, can help me here as well?

FRidh commented on pkgs/applications/office/fava/default.nix:
```diff
@@ -0,0 +1,73 @@
+
+    buildInputs = with python3Packages; [ nose ];
+
+    # Automatic tests cannot be run because it needs to import some local modules for tests.
```
what modules? Its quite common that we have some test dependencies.

FRidh commented on pkgs/applications/office/fava/default.nix:
```diff
@@ -0,0 +1,73 @@
+  # depends on beancount-2.0b2, but beancount as in pkgs.beancount is from the
+  # master branch of the beancount repository (as the maintainer does not
+  # release versions).
+  beancount = python3Packages.buildPythonPackage rec {
```
application, or library? In case of the former `buildPythonApplication`.

matthiasbeyer replied:
In this case an application, IMO, as it is required as dependency of `fava` - not completely sure, though.

FRidh commented on pkgs/applications/office/fava/default.nix:
```diff
@@ -0,0 +1,73 @@
+  propagatedBuildInputs = with python3Packages;
+    [ flask dateutil pygments livereload wheel markdown2 flaskbabel tornado
+      click
+      beancount-pygments-lexer ] ++
```
unnecessary newline and ++

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1870,6 +1870,24 @@ in modules // {
+      sha256 = "0qil39ckf6jhlgjwlxjrfphmx5zs9zhp0b92mrw99kbi06f1qcdq";
+    };
+
+    buildInputs = with self; [ pygments ];
```
are you sure this is not a runtime dependency?

matthiasbeyer replied:
Nope, I'm not. I would need to change it to `propagatedBuildInputs`, am I right?

FRidh replied:
if it is a runtime dependency, then yes.

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -15374,6 +15392,25 @@ in modules // {
+      sha256 = "0m8b4f6wkrsh45ldqrq2l9ynv2dzlgb34bkwc7hzi0cihsfgwaca";
+    };
+
+    buildInputs = with self; [ tornado six ];
```
are you sure these are not runtime dependencies?

Mic92 commented:
javascript/css must be compiled with nodejs. Fortunately this is already done in the release tarball.
So you just have to switch the source:

```patch
diff --git a/pkgs/applications/office/fava/default.nix b/pkgs/applications/office/fava/default.nix
index b87d86a..067c48d 100644
--- a/pkgs/applications/office/fava/default.nix
+++ b/pkgs/applications/office/fava/default.nix
@@ -53,8 +53,8 @@ python3Packages.buildPythonApplication rec {
   name = "fava-${version}";
 
   src = fetchurl {
-    url = "https://github.com/aumayr/fava/archive/v${version}.tar.gz";
-    sha256 = "0jj4prv9f8h4zzynw006gvlg5m7ccwhrgrgdlklncpvb1rkarzjc";
+    url = "https://github.com/aumayr/fava/releases/download/v${version}/beancount-fava-${version}.tar.gz";
+    sha256 = "1a5ws0amy2wf9vh4rxk9vn8822zfyizfprhrlnndwfps6mxd63np";
   };
 
   propagatedBuildInputs = with python3Packages;
```

matthiasbeyer commented:
Awesome! :tada: What @Mic92 suggested worked for me. Now this seems to work!

Any more things to clean up before squashing my fixup commits?

Mic92 commented:
Fava used `py.test` for testing. Add  `pythonPackages.pytest` to buildInputs and

```
checkPhase = "py.test";
```

matthiasbeyer committed:
Add pytest setup for fava

matthiasbeyer committed:
fava: Disable tests

matthiasbeyer commented:
With disabled tests, this works for me.... I don't know why the tests do not work ... I'm not sure we need them anyways :-P

FRidh commented:
>  I'm not sure we need them anyways :-P

to make sure the package actually works

matthiasbeyer committed:
fava: 0.3.0 -> 1.0

Also updates the fava-only dependency of beancount: 2.0b11 -> 2.0b12

Beancount itself does not have releases, they are just published as such
on pypi, therefor we have a special package for fava here.

Mic92 commented:
The problem is, that the tests are not included in the release tarball, which on the other hand includes generated css/javascript.

matthiasbeyer committed:
fava: Remove deps: livereload beancount-pygments-lexer

matthiasbeyer commented:
@Mic92 do you know a way to fix this for the tests?

Mic92 commented:
```patch
diff --git a/pkgs/applications/office/fava/default.nix b/pkgs/applications/office/fava/default.nix
index 2d20b17..d7fbfce 100644
--- a/pkgs/applications/office/fava/default.nix
+++ b/pkgs/applications/office/fava/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, pkgs, fetchurl, python3Packages }:
+{ stdenv, pkgs, fetchurl, python3Packages, fetchFromGitHub, fetchzip, python3 }:
 
 let
 
@@ -52,15 +52,34 @@ python3Packages.buildPythonApplication rec {
   version = "1.0";
   name = "fava-${version}";
 
-  src = fetchurl {
+  src = fetchFromGitHub {
+    owner = "aumayr";
+    repo = "fava";
+    rev = "v${version}";
+    sha256 = "0dm4x6z80m04r9qa55psvz7f41qnh13hnj2qhvxkrk22yqmkqrka";
+  };
+
+  # css/js is only included in the release build.
+  assets = fetchzip {
     url = "https://github.com/aumayr/fava/releases/download/v${version}/beancount-fava-${version}.tar.gz";
-    sha256 = "1yks1vy4cskv1h4dib0zx2cl8xz4kz4qx9n8v76w3di21w2gmak3";
+    sha256 = "1vvidwfn5882dslz6qqkkd84m7w52kd34x10qph8yhipyjv1dimc";
   };
 
-  buildInputs = with python3Packages; [ pytest ];
+  #doCheck = false;
+  buildInputs = with python3Packages; [ pytest_30 ];
+
+  checkPhase = ''
+    # pyexcel is optional
+    # the other 2 tests fail due non-unicode locales
+    PATH=$out/bin:$PATH pytest tests \
+      --ignore tests/test_util_excel.py \
+      --ignore tests/test_cli.py \
+      --ignore tests/test_translations.py \
+  '';
 
-  doCheck = false;
-  checkPhase = "py.test";
+  postInstall = ''
+    cp -r $assets/fava/static/gen $out/${python3.sitePackages}/fava/static
+  '';
 
   propagatedBuildInputs = with python3Packages;
     [ flask dateutil pygments wheel markdown2 flaskbabel tornado
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index 527a2b8..e6f18c1 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -4812,6 +4812,16 @@ in modules // {
     };
   };
 
+  pytest_30 = self.pytest_27.override rec {
+    name = "pytest-3.0.3";
+
+    propagatedBuildInputs = with self; [ hypothesis py ];
+    src = pkgs.fetchurl {
+      url = "mirror://pypi/p/pytest/${name}.tar.gz";
+      sha256 = "1rxydacrdb8s312l3bn0ybrqsjp13abzyim1x21s80386l5504zj";
+    };
+  };
+
   pytestcache = buildPythonPackage rec {
     name = "pytest-cache-1.0";
     src = pkgs.fetchurl {
```

matthiasbeyer committed:
pytest_30: Add override from pytest_27

matthiasbeyer committed:
fava: Add checkphase override, assets

As suggested by @Mic92 in
https://github.com/NixOS/nixpkgs/pull/15785#issuecomment-255036164

matthiasbeyer commented:
@Mic92 Added your suggestions.
